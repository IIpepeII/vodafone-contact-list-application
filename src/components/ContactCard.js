/*
* Contact item for Dashboard contacts grid.
* props:
* - data: individual contact object from contacts state getting in Dashboard component.
*/
import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';

const ContactCard = (props) => {
  const {
    id,
    firstName,
    lastName,
    email,
    cell,
  } = props.data;

  return (
    <div className="contact-card col-sm-6 col-md-4 text-center">
      <Link to={`/contact/${id}`} className="contact-card__clickwrapper">
        <header className="contact-card__header py-3">
          { lastName }, { firstName }
        </header>
        <div className="contact-card__content">
          <small className="pt-2">{ email }</small>
          <small className="py-2">{ cell }</small>
        </div>
      </Link>
    </div>
  );
};

ContactCard.propTypes = {
  data: PropTypes.objectOf(PropTypes.string).isRequired,
};

export default ContactCard;
