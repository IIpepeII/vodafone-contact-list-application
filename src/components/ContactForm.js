/*
* Form for adding and editing contact data.
* Used in ./AddContactPage.js and EditContactPage.js
* props:
* - contact: actual contact data if using the form for editing contact in <EditContactPage />.
* - page: needed to properly render text on form button. 'Save contact' or 'Add contact'.
* - getFormData: function for sending data to parent. <AddContactPage /> or <EditContactPage />
*/
import React, { Component } from 'react';
import {
  Button,
  Form,
  FormGroup,
  Label,
  Input,
} from 'reactstrap';
import PropTypes from 'prop-types';

class ContactForm extends Component {
  static propTypes = {
    contact: PropTypes.objectOf(PropTypes.string),
    page: PropTypes.string.isRequired,
    getFormData: PropTypes.func.isRequired,
  }

  // If form is child of <AddContactPage /> there is no contact prop.
  static defaultProps = {
    contact: {
      firstName: '',
      lastName: '',
      email: '',
      cell: '',
    },
  };

  state = {
    contact: {
      firstName: this.props.contact.firstName,
      lastName: this.props.contact.lastName,
      email: this.props.contact.email,
      cell: this.props.contact.cell,
    },
    error: false,
  }

  // By onchanging form fields update local state by field name property.
  updateContactState = (e) => {
    const { name, value } = e.target;
    this.setState(prevState => ({
      contact: {
        ...prevState.contact,
        [name]: value,
      },
    }));
  }

  // Handle submit form
  handleClickSaveContact = (e) => {
    e.preventDefault();

    const {
      firstName,
      lastName,
      email,
      cell,
    } = this.state.contact;

    /*
    * Regex for email validation.
    */
    const emailReg = /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/;

    // Form validation. Check if any field is empty and email address is valid based on regex
    if (firstName.trim().length === 0
      || lastName.trim().length === 0
      || email.trim().length === 0
      || cell.trim().length === 0) {
      this.setState({ error: 'Please, fill out every field!' });
    } else if (!emailReg.test(email)) {
      this.setState({ error: 'Please, enter a valid email address!' });
    } else {
      this.props.getFormData(this.state.contact);
    }
  }

  render() {
    return (
      <Form onSubmit={this.handleClickSaveContact}>
        <FormGroup>
          <Label for="firstName">First name:</Label>
          <Input
            type="text"
            value={this.state.contact.firstName}
            name="firstName"
            placeholder="Enter First Name"
            onChange={this.updateContactState}
          />
        </FormGroup>
        <FormGroup>
          <Label for="lastName">Last name:</Label>
          <Input
            type="text"
            value={this.state.contact.lastName}
            name="lastName"
            placeholder="Enter Last Name"
            onChange={this.updateContactState}
          />
        </FormGroup>
        <FormGroup>
          <Label for="lastName">E-mail:</Label>
          <Input
            type="text"
            value={this.state.contact.email}
            name="email"
            placeholder="Enter Email"
            onChange={this.updateContactState}
          />
        </FormGroup>
        <FormGroup>
          <Label for="cell">Phone:</Label>
          <Input
            type="text"
            value={this.state.contact.cell}
            name="cell"
            placeholder="Enter Phone"
            onChange={this.updateContactState}
          />
        </FormGroup>
        {
          this.state.error && <p className="error">{this.state.error}</p>
        }
        <div className="text-center pt-4">
          <Button className="btn btn--empty btn--gray-empty">
            {this.props.page === 'editContact' ? 'Save Contact' : 'Add Contact'}
          </Button>
        </div>
      </Form>
    );
  }
}

export default ContactForm;
