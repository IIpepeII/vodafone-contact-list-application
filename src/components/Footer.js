/*
* Footer for the app.
* Component is wrapped in withRouter HOC to have access to React Router's history object.
* History object is needed to get actual pathname.
* Based on actual path we display or hide add new contact button.
*/
import React from 'react';
import { Link, withRouter } from 'react-router-dom';

const Footer = ({ history }) => (
  <footer className="footer text-center py-5">
    All Rights Reserved VSSB 2018.
    {/* On addcontact page do not display add new button */}
    {
      !(history.location.pathname === '/addcontact')
      && (
      <Link
        className="btn btn--red btn--rounded btn--addnew"
        to="/addcontact"
      >
        &#43;
      </Link>
      )
    }
  </footer>
);

export default withRouter(Footer);
