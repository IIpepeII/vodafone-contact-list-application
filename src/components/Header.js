/*
* Header of the app.
* Props:
* - dispatchTextFilter: it is for sending filter text from search input to store via setTextFilter action creator
*/
import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import { FormGroup, Label, Input } from 'reactstrap';
import setTextFilter from '../actions/filters';
import logo from '../images/vodafone_logo.svg';

export const Header = (props) => {
  const { dispatchTextFilter } = props;

  return (
    <header className="header container py-4">
      <div className="header__logo">
        <Link to="/"><img src={logo} alt="Vodafone" /></Link>
      </div>
      <div className="header__search d-none d-sm-block">
        <FormGroup>
          <Label for="search" className="sr-only">Search for contact</Label>
          <Input
            type="text"
            placeholder="Search for contact"
            onChange={e => dispatchTextFilter(e.target.value)}
          />
        </FormGroup>
      </div>
    </header>
  );
};

Header.propTypes = {
  dispatchTextFilter: PropTypes.func.isRequired,
};

/*
* Create dispatchTextFilter function to link setTextFilter action creator with the component via props.
*/
const mapDispatchToProps = dispatch => ({
  dispatchTextFilter: text => dispatch(setTextFilter(text)),
});

export default connect(null, mapDispatchToProps)(Header);
