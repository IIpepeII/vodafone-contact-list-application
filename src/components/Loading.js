/*
* Loading image
* Used in ./Dashboard.js while contactsAreLoading Redux state is true.
*/
import React from 'react';
import loader from '../images/loading.svg';

const Loading = () => (
  <div className="loading">
    <img src={loader} alt="loading..." />
  </div>
);

export default Loading;
