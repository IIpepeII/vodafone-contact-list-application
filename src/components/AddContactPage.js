/*
* Add new contact component. I use shortid package to generate id.
* The form is in ./ContactForm.js <ContactForm />
* Props:
* - dispatchAddContact for sending data to the store via addContact action creator.
* - React Router history object for dropping user back to dashboard after saving the form.
*/
import React, { Component } from 'react';
import { connect } from 'react-redux';
import shortid from 'shortid';
import PropTypes from 'prop-types';
import { addContact } from '../actions/contacts';
import ContactForm from './ContactForm';

class AddContactPage extends Component {
  static propTypes = {
    dispatchAddContact: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
  }

  // Create contact object then sending data to the store then drop to dashboard.
  handleAddContact = (data) => {
    const contact = {
      id: shortid.generate(),
      ...data,
    };

    this.props.dispatchAddContact(contact);
    this.props.history.push('/');
  }

  render() {
    return (
      <div>
        <h1>Add New Contact</h1>
        {/* ContactForm component render text on form button based on page prop. */}
        <ContactForm
          getFormData={this.handleAddContact}
          page="addContact"
        />
      </div>
    );
  }
}

/*
* Create dispatchAddContact function to link addContact action creator with the component via props.
*/
const mapDispatchToProps = dispatch => ({
  dispatchAddContact: contact => dispatch(addContact(contact)),
});

export default connect(null, mapDispatchToProps)(AddContactPage);
