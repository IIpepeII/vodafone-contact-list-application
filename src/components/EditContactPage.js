/*
* Edit contact component for update contact data.
* The form is in ./ContactForm.js <ContactForm />
* Props:
* - contact: actual contact data from contacts state by id in URL.
* - dispatchEditContact: it is for sending data to the store via editContact action creator.
* - React Router history for dropping user back to dashboard after saving the form.
*/
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { editContact } from '../actions/contacts';
import ContactForm from './ContactForm';

class EditContactPage extends Component {
  static propTypes = {
    contact: PropTypes.objectOf(PropTypes.string).isRequired,
    dispatchEditContact: PropTypes.func.isRequired,
    history: PropTypes.object.isRequired,
  }

  /*
  * Handle form submit.
  * Call dispatchEditContact to send data to store via editContact action creator then drop user to dashboard
  */
  handleEditContact = (data) => {
    this.props.dispatchEditContact(this.props.contact.id, data);
    this.props.history.push('/');
  }

  render() {
    const {
      firstName,
      lastName,
    } = this.props.contact;
    return (
      <div>
        <h1>{firstName} {lastName}&apos;s Profile</h1>
        <div className="edit-contact-page px-lg-7 pb-4">
          <ContactForm
            contact={this.props.contact}
            getFormData={this.handleEditContact}
            page="editContact"
          />
        </div>
      </div>
    );
  }
}

/*
* Based on React Router's match object find contact in contacts state array by id.
*/
const mapStateToProps = (state, props) => ({
  contact: state.contacts.find(contact => contact.id === props.match.params.id),
});

/*
* Create dispatchAEdiContact function to link editContact action creator with the component via props.
*/
const mapDispatchToProps = dispatch => ({
  dispatchEditContact: (id, updatedData) => dispatch(editContact(id, updatedData)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditContactPage);
