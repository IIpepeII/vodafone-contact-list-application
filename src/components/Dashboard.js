/*
* Dashboard component. It is the initial page.
* Props:
* - filteredContacts: contacts state from store filtered with function in selectors/contacts.js
* - contacts: unfiltered contacts state from store. Needed to display letters for filter
* - filter: filter state to add active class to filter letters
* - loading: loading state to decide whether display <Loading /> or contact list
* - dispatchTextFilter: send filter text to store via setTextFilter action creator
*/
import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import filterContacts from '../selectors/contacts';
import ContactCard from './ContactCard';
import Loading from './Loading';
import setTextFilter from '../actions/filters';

export class Dashboard extends Component {
  static propTypes = {
    filteredContacts: PropTypes.arrayOf(PropTypes.object),
    contacts: PropTypes.arrayOf(PropTypes.object),
    filter: PropTypes.string,
    loading: PropTypes.bool.isRequired,
    dispatchTextFilter: PropTypes.func.isRequired,
  }

  static defaultProps = {
    filteredContacts: [],
    contacts: [],
    filter: '',
  }

  // Method for getting individual first letters from last names for filtering
  getFirstLettersFromLastNames = () => {
    const letters = this.props.contacts
      .map(contact => contact.lastName.charAt(0))
      .sort();
    return [...new Set(letters)];
  }

  render() {
    const {
      filteredContacts,
      dispatchTextFilter,
      filter,
    } = this.props;

    return (
      <div className="dashboard">
        <h1 className="text-center">Contacts</h1>
        <div className="text-center pt-3">
          {
            this.props.contacts.length > 0
            && this.getFirstLettersFromLastNames().map(letter => (
              <button
                type="button"
                className={`text-uppercase btn--filter mx-2 ${filter === letter ? 'active' : ''}`}
                id={`filter-${letter}`}
                key={letter}
                onClick={() => dispatchTextFilter(letter)}
              >
                { letter }
              </button>
            ))
          }
        </div>
        {/* Display 'Clear filter' button if contacts are filtered */}
        {
          filter.length > 0 && (
          <div className="text-center py-3">
            <button
              type="button"
              className="btn btn--empty btn--red-empty"
              onClick={() => dispatchTextFilter('')}
            >
            Clear filter
            </button>
          </div>
          )
        }
        {/* If contactsAreLoading state is true display <Loading /> else display contacts */}
        <div className="row mt-5">
          {
            this.props.loading
              ? <Loading />
              : filteredContacts.map(contact => <ContactCard data={contact} key={contact.id} />)
          }
        </div>
      </div>
    );
  }
}

/*
* Get states from the Redux store to add as props to the component
*/
const mapStateToProps = state => ({
  filteredContacts: filterContacts(state.contacts, state.filters.text),
  contacts: state.contacts,
  filter: state.filters.text,
  loading: state.contactsAreLoading,
});

/*
* Create dispatchTextFilter function to link setTextFilter action creator with the component via props
*/
const mapDispatchToProps = dispatch => ({
  dispatchTextFilter: text => dispatch(setTextFilter(text)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Dashboard);
