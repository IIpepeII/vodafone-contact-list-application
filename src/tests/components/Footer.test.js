import React from 'react';
import { shallow } from 'enzyme';
import { Footer } from '../../components/Footer';

test('should render Footer correctly if current page is NOT addcontact', () => {
  const history = {
    location: {
      pathname: '/',
    }
  }
  const wrapper = shallow(<Footer history={history} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render Footer correctly if current page IS addcontact', () => {
  const history = {
    location: {
      pathname: '/addcontact',
    }
  }
  const wrapper = shallow(<Footer history={history} />);
  expect(wrapper).toMatchSnapshot();
});
