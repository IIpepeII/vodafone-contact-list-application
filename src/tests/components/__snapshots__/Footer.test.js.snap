// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render Footer correctly if current page IS addcontact 1`] = `
<footer
  className="footer text-center py-5"
>
  All Rights Reserved VSSB 2018.
</footer>
`;

exports[`should render Footer correctly if current page is NOT addcontact 1`] = `
<footer
  className="footer text-center py-5"
>
  All Rights Reserved VSSB 2018.
  <Link
    className="btn btn--red btn--rounded btn--addnew"
    replace={false}
    to="/addcontact"
  >
    +
  </Link>
</footer>
`;
